apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,158 @@
+ * object. If there were no validation errors the result will contain an empty array
+ * of errors.
+ *
+ * The provided {@link PCDCollection} is not checked unless either this function
+ * determines the user is logged in or the {@link forceCheckPCDs} argument is `true`.
+ */
+function validateAppState(
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): ValidationErrors {
+  const validationErrors: string[] = [];
+
+  const loggedOut = !self;
+
+  const identityPCDFromCollection = pcds?.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )?.[0] as SemaphoreIdentityPCD | undefined;
+
+  if (forceCheckPCDs || !loggedOut) {
+    if (!identityPCDFromCollection) {
+      validationErrors.push(
+        "'pcds' field in app state does not contain an identity PCD"
+      );
+    }
+  }
+
+  if (loggedOut) {
+    return {
+      errors: validationErrors,
+      userUUID: undefined
+    };
+  }
+
+  if (!self) {
+    validationErrors.push("missing 'self'");
+  }
+
+  if (!identity) {
+    validationErrors.push("missing 'identity'");
+  }
+
+  if (!pcds) {
+    validationErrors.push("missing 'pcds'");
+  }
+
+  if (pcds?.size() === 0) {
+    validationErrors.push("'pcds' contains no pcds");
+  }
+
+  const identityFromPCDCollection = identityPCDFromCollection?.claim?.identity;
+  const commitmentOfIdentityPCDInCollection =
+    identityFromPCDCollection?.commitment?.toString();
+  const commitmentFromSelfField = self?.commitment;
+  const commitmentFromIdentityField = identity?.commitment?.toString();
+
+  if (commitmentOfIdentityPCDInCollection !== commitmentFromSelfField) {
+    validationErrors.push(
+      `commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})` +
+        ` does not match commitment in 'self' field of app state (${commitmentFromSelfField})`
+    );
+  }
+
+  if (commitmentFromSelfField !== commitmentFromIdentityField) {
+    validationErrors.push(
+      `commitment in 'self' field of app state (${commitmentFromSelfField})` +
+        ` does not match commitment of 'identity' field of app state (${commitmentFromIdentityField})`
+    );
+  }
+
+  if (commitmentFromIdentityField !== commitmentOfIdentityPCDInCollection) {
+    validationErrors.push(
+      `commitment of 'identity' field of app state (${commitmentFromIdentityField})` +
+        ` does not match commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})`
+    );
+  }
+
+  return {
+    errors: validationErrors,
+    userUUID: self?.uuid
+  };
+}
+
+/**
+ * Logs validation errors to the console, and uploads them to the server so that
+ * we have records and are able to identify common types of errors. Does not leak
+ * sensitive information, such as decrypted versions of e2ee storage.
+ */
+async function logValidationErrors(errors: ValidationErrors): Promise<void> {
+  try {
+    const user = loadSelf();
```

Comment:
It's interesting that this might be a different self than the one passed in to the validator, which seems confusing.  Consider passing that argument through to this function, and only calling loadSelf() as a backup option (if at all).

Reply:
Still needs a response.

Reply:
This covers the case where the validator doesn't have access to a user because of the place it was called, but there is actually a user stored on the page. eg. this can happen in the localstorage validator case.